src/read.rs

```diff
@@ -226,48 +225,45 @@ impl<R> Reader<R>
     /// The format is determined by attempting to construct each specific format reader until one
     /// is successful.
     pub fn new(mut reader: R) -> Result<Self, ReadError> {
-        let mut reader_errors = Vec::new();
 
         #[cfg(feature="wav")]
         {
-            let maybe_err = match hound::WavReader::new(&mut reader) {
-                Ok(_) => None,
-                Err(err) => Some(err),
+            let is_wav = match hound::WavReader::new(&mut reader) {
+                Err(hound::Error::FormatError(_)) => false,
+                _ => true,
```

Comment:
> Oh right, we construct them twice.... Maybe it would be a good idea to re-use it?

The reason we have to construct it twice is due to an issue with ownership - I mention this in the parent comment:

> Also, this method requires that the successfully detected inner format
> reader is constructed *twice*:
> 
> 1. By passing the reader argument via `&mut` to test for the format. We
> must use `&mut` to avoid moving the reader argument into the format
> reader's constructor, otherwise we would not be able to re-use the
> reader argument when testing the following formats.
> 2. Passing the reader argument via move for constructing the `Reader`
> that will be returned after successfully detecting the format.

> If we ignore IO errors and retry multiple times with different formats this means a much longer period to wait

True, this is a fair point - I'll add an extra branch that ensures we return immediately if we get any errors other than format-related errors.